File: packages/senri_ffi_js/src/globals.ts
/*
 * Copyright (c) 2026 TT23XR Studio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { FFIAdapter } from './types/adapter';
import { FFIError } from './errors';
import { getResourceRegistry } from './resource-registry';

let _adapter: FFIAdapter | null = null;
let _version = 0;

export function setGlobalAdapter(adapter: FFIAdapter, force?: boolean): void {
  if (_adapter && _adapter !== adapter) {
    const registry = getResourceRegistry();
    if (registry.hasActiveResources()) {
      if (!force) {
        const memCount = registry.getActiveMemoryCount();
        const cbCount = registry.getActiveCallbackCount();
        throw new FFIError(
          'Cannot switch FFI backend: there are ' + memCount + ' active memory allocation(s) and ' +
          cbCount + ' active callback(s). Free all resources before switching, ' +
          'or pass `true` as the second argument to setGlobalAdapter to force switch (risky).'
        );
      }
      if (typeof console !== 'undefined' && console.warn) {
        console.warn(
          '[SenRi FFI] WARNING: Force-switching adapter with active resources. ' +
          'Existing pointers and callbacks may cause crashes or memory leaks.'
        );
      }
      registry.clear();
    }
  }
  _adapter = adapter;
  _version++;
}

export function getGlobalAdapter(): FFIAdapter {
  if (!_adapter) throw new FFIError('Adapter not initialized');
  return _adapter;
}

export function getGlobalAdapterVersion(): number {
  return _version;
}

export function isAdapterInitialized(): boolean {
  return _adapter !== null;
}

File: packages/senri_ffi_py/src/senri_ffi/globals.py
from __future__ import annotations

import sys

from senri_ffi.errors import FFIError
from senri_ffi.registry import get_resource_registry
from senri_ffi.types.adapter import FFIAdapter

_adapter: FFIAdapter | None = None
_version: int = 0


def set_global_adapter(adapter: FFIAdapter, force: bool = False) -> None:
    global _adapter, _version
    if _adapter is not None and _adapter is not adapter:
        registry = get_resource_registry()
        if registry.has_active_resources():
            if not force:
                mem_count = registry.get_active_memory_count()
                cb_count = registry.get_active_callback_count()
                raise FFIError(
                    f"Cannot switch FFI backend: there are {mem_count} active memory allocation(s) and "
                    f"{cb_count} active callback(s). Free all resources before switching, "
                    "or pass force=True to set_global_adapter to force switch (risky)."
                )
            print(
                "[SenRi FFI] WARNING: Force-switching adapter with active resources. "
                "Existing pointers and callbacks may cause crashes or memory leaks.",
                file=sys.stderr,
            )
            registry.clear()
    _adapter = adapter
    _version += 1


def get_global_adapter() -> FFIAdapter:
    if _adapter is None:
        raise FFIError("Adapter not initialized")
    return _adapter


def get_global_adapter_version() -> int:
    return _version


def is_adapter_initialized() -> bool:
    return _adapter is not None
